### PR TITLE
RandomGeoSampler: fix performance regression

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes a bug I introduced all the way back in #477. Thanks @yichiac for noticing this!

We want to try to sample from a pixel-aligned grid whenever possible to avoid resampling in the average case. With the changes in this PR, we should revert back to the behavior we documented in our paper, where preprocessing greatly improves sampling performance. GridGeoSampler was not affected by this bug.

I still want to create an I/O benchmarking script similar to what we used in our original paper to ensure that this indeed works as advertised, so will keep as a draft for now.

### CODE REVIEW

File: torchgeo/samplers/utils.py [hunk 1]
```diff
@@ -55,17 +55,16 @@ def get_random_bounding_box(
     """
     t_size = _to_tuple(size)
 
-    width = (bounds.maxx - bounds.minx - t_size[1]) // res
-    height = (bounds.maxy - bounds.miny - t_size[0]) // res
+    # May be negative if bounding box is smaller than patch size
+    width = (bounds.maxx - bounds.minx - t_size[1]) / res
+    height = (bounds.maxy - bounds.miny - t_size[0]) / res
```
Comment (adamjstewart) [inline torchgeo/samplers/utils.py:60]: This is the width/height in pixel units. It no longer needs to be an integer, float is fine too. We cast to integer elsewhere.

File: torchgeo/samplers/utils.py [hunk 2]
```diff
@@ -55,17 +55,16 @@ def get_random_bounding_box(
...
 
     minx = bounds.minx
     miny = bounds.miny
 
-    # random.randrange crashes for inputs <= 0
```
Comment (adamjstewart) [inline torchgeo/samplers/utils.py:64]: This is a dead comment, we no longer use `random.randrange` and the input can no longer be negative

File: torchgeo/samplers/utils.py [hunk 3]
```diff
@@ -55,17 +55,16 @@ def get_random_bounding_box(
...
-    if width > 0:
```
Comment (adamjstewart) [inline torchgeo/samplers/utils.py:65]: No longer need to guard for negative numbers, it doesn't matter if the sample starts outside the bounds of the image for bounding boxes smaller than the patch size.

File: torchgeo/samplers/utils.py [hunk 4]
```diff
@@ -55,17 +55,16 @@ def get_random_bounding_box(
...
-        minx += torch.rand(1).item() * width * res
-    if height > 0:
-        miny += torch.rand(1).item() * height * res
+    # Use an integer multiple of res to avoid resampling
+    minx += int(torch.rand(1).item() * width) * res
```
Comment (adamjstewart) [inline torchgeo/samplers/utils.py:66]: This is the only real important line

**Comment by adamjstewart** [comment]
Evaluation using #1972:

| | raw (random) | raw (grid) | preprocessed (random) | preprocessed (grid) |
| --- | --- | --- | --- | --- |
| before | 17.223 | 10.974 | 15.685 | 4.6075 |
| after | 17.360 | 11.032 | 9.6127 | 4.6673 |

A 60% speedup for RandomGeoSampler when data is preprocessed (same CRS, res, TAP, COG).

**Review by isaaccorley — APPROVED** [review]

**Pull request merged**